volttron/platform/auth.py

```diff
@@ -131,6 +147,196 @@ def setup_zap(self, sender, **kwargs):
+                        try:
+                            self.vip.rpc.call(
+                                entry.identity, "set_rpc_authorizations",
+                                method_str=method, capabilities=entry.rpc_method_authorizations[method]).get(timeout=4)
+                        except gevent.Timeout:
+                            _log.error(f"{entry.identity} "
+                                       f"has timed out while attempting to update rpc_method_authorizations")
+                        except RemoteError:
+                            _log.error(f"Method {method} does not exist.")
+
+    def update_auth_file_rpc_method_authorizations(self, entries):
+        """
+        Update rpc_method_authorizations field for each auth entry in the auth file.
+        :param entries: Entries read in from the auth file
+        :return: An updated list of entries to update the auth file
+        """
+        new_entries = copy.deepcopy(entries)
+        for entry in new_entries:
+            if entry.identity is not None and \
+                entry.identity not in PROCESS_IDENTITIES and \
+                entry.identity is not CONTROL_CONNECTION:
+                rpc_method_authorizations = self.get_entry_rpc_method_authorizations(entry.identity)
+                for method in rpc_method_authorizations:
+                    # Check if the rpc method exists in the auth file entry
+                    if method not in entry.rpc_method_authorizations:
+                        if rpc_method_authorizations[method] is not None:
+                            # Create it and set it to have the rpc capabilities provided by the agent if it is available
+                            # If it is not available, this will be set when the agent starts/connects to the platform
+                            entry.rpc_method_authorizations[method] = rpc_method_authorizations[method]
+                        else:
+                            entry.rpc_method_authorizations[method] = []
+                    # Check if the rpc method does not have any rpc capabilities
+                    if not entry.rpc_method_authorizations[method]:
+                        if rpc_method_authorizations[method] is not None:
+                            # Set it to have the rpc capabilities provided by the agent if it is available
+                            entry.rpc_method_authorizations[method] = rpc_method_authorizations[method]
+                        else:
+                            entry.rpc_method_authorizations[method] = []
+                    # Check if the rpc method's capabilities match what have been provided and the agent was reachable
+                    if entry.rpc_method_authorizations[method] != rpc_method_authorizations[method] and rpc_method_authorizations[method] is not None:
+                        try:
+                            self.vip.rpc.call(
+                                entry.identity, "set_rpc_authorizations",
```

Comment:
Can we collect all the methods that need new authorizations and sent it to agent and let subsystem handle it?

Reply:
Fixed.